### ISSUE TITLE
Correctness Issue with Torch FSDP2

### ISSUE THREAD

**Issue by haok1402** [NVIDIA/Megatron-LM#2769]
Hi, I'm using `core_v0.15.0` release with Megatron-LM and running a mid-training from the Llama3-8B checkpoint. 

Exact same configuration, using the Megatron's `use-distributed-optimizer`, the loss starts off at 2.404998E+00, which is a correct range for a pretrained model.

```
 [2025-12-28 00:20:24] iteration        1/     100 | consumed samples:         1024 | elapsed time per iteration (ms): 27003.9 | throughput per GPU (TFLOP/s/GPU): 452.7 | learning rate: 9.900000E-05 | global batch size:  1024 | lm loss: 2.404998E+00 | loss scale: 1.0 | grad norm: 3.048 | number of skipped iterations:   0 | number of nan iterations:   0 |Number of parameters in transformer block in billions:  6.98 
 [2025-12-28 00:20:49] iteration        2/     100 | consumed samples:         2048 | elapsed time per iteration (ms): 24621.6 | throughput per GPU (TFLOP/s/GPU): 496.6 | learning rate: 9.800000E-05 | global batch size:  1024 | lm loss: 4.466884E+00 | loss scale: 1.0 | grad norm: 44.051 | number of skipped iterations:   0 | number of nan iterations:   0 |
```

However, when I try to use torch FSDP2 via `use-torch-fsdp2`, the loss starts off with at 1.386388E+01, which isn't a correct range...
 
```
 [2025-12-28 00:23:52] iteration        1/     100 | consumed samples:         1024 | elapsed time per iteration (ms): 37835.9 | throughput per GPU (TFLOP/s/GPU): 323.1 | learning rate: 9.900000E-05 | global batch size:  1024 | lm loss: 1.386388E+01 | loss scale: 1.0 | grad norm: 346.039 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-12-28 00:24:26] iteration        2/     100 | consumed samples:         2048 | elapsed time per iteration (ms): 34441.6 | throughput per GPU (TFLOP/s/GPU): 355.0 | learning rate: 9.800000E-05 | global batch size:  1024 | lm loss: 3.159755E+01 | loss scale: 1.0 | grad norm: 544.112 | number of skipped iterations:   0 | number of nan iterations:   0 |
```

The full command line arguments are attached below for reference.

```bash
#!/bin/bash

declare -A MODEL_CONFIG
MODEL_CONFIG[tokenizer-type]=HuggingFaceTokenizer
MODEL_CONFIG[tokenizer-model]=meta-llama/Meta-Llama-3-8B
MODEL_CONFIG[untie-embeddings-and-output-weights]=true
MODEL_CONFIG[position-embedding-type]=rope
MODEL_CONFIG[rotary-base]=500000
MODEL_CONFIG[max-position-embeddings]=8192
MODEL_CONFIG[num-layers]=32
MODEL_CONFIG[hidden-size]=4096
MODEL_CONFIG[ffn-hidden-size]=14336
MODEL_CONFIG[hidden-dropout]=0.0
MODEL_CONFIG[disable-bias-linear]=true
MODEL_CONFIG[swiglu]=true
MODEL_CONFIG[group-query-attention]=true
MODEL_CONFIG[num-attention-heads]=32
MODEL_CONFIG[num-query-groups]=8
MODEL_CONFIG[kv-channels]=128
MODEL_CONFIG[attention-dropout]=0.0
MODEL_CONFIG[init-method-std]=0.02
MODEL_CONFIG[normalization]=RMSNorm
MODEL_CONFIG[norm-epsilon]=1e-5

declare -A TRAIN_CONFIG
TRAIN_CONFIG[lr]=1e-4
TRAIN_CONFIG[optimizer]=adam
TRAIN_CONFIG[adam-beta1]=0.9
TRAIN_CONFIG[adam-beta2]=0.95
TRAIN_CONFIG[weight-decay]=0.1
TRAIN_CONFIG[clip-grad]=1.0
TRAIN_CONFIG[train-iters]=100
TRAIN_CONFIG[micro-batch-size]=1
TRAIN_CONFIG[global-batch-size]=1024
TRAIN_CONFIG[seq-length]=2048
TRAIN_CONFIG[bf16]=true

# TRAIN_CONFIG[use-distributed-optimizer]=true
TRAIN_CONFIG[use-torch-fsdp2]=true
TRAIN_CONFIG[no-gradient-accumulation-fusion]=true

TRAIN_CONFIG[log-interval]=1
TRAIN_CONFIG[no-one-logger]=true
TRAIN_CONFIG[log-throughput]=true
TRAIN_CONFIG[pretrained-checkpoint]=checkpoints/llama3-8b/megatron/
TRAIN_CONFIG[load]=checkpoints/debug/megatron/
TRAIN_CONFIG[save]=checkpoints/debug/megatron/
TRAIN_CONFIG[per-split-data-args-path]=checkpoints/debug/per-split-data-args.json
TRAIN_CONFIG[save-interval]=125
TRAIN_CONFIG[eval-iters]=0

MAIN_ARGS=()
for key in ${!MODEL_CONFIG[@]}; do val=${MODEL_CONFIG[$key]}; [[ $val == true ]] && MAIN_ARGS+=(--$key) || MAIN_ARGS+=(--$key $val); done
for key in ${!TRAIN_CONFIG[@]}; do val=${TRAIN_CONFIG[$key]}; [[ $val == true ]] && MAIN_ARGS+=(--$key) || MAIN_ARGS+=(--$key $val); done

TRUN_ARGS=()
TRUN_ARGS+=(--nnodes=1 --nproc-per-node=8)
TRUN_ARGS+=(--rdzv-backend=c10d --rdzv-endpoint=localhost:15213)

SCRIPT=Megatron-LM/pretrain_gpt.py
STDOUT=checkpoints/debug/midtrain.log

mkdir -p $(dirname $STDOUT)
export WANDB_RUN_ID=$(echo ${TRAIN_CONFIG[WANDB_EXP_NAME]} | md5sum | cut -c1-5)
torchrun ${TRUN_ARGS[@]} $SCRIPT ${MAIN_ARGS[@]} 2>&1 | tee $STDOUT
```

**Comment by haok1402** [comment]
This does work with `use-megatron-fsdp`.

```
 [2025-12-28 01:33:27] iteration        1/     100 | consumed samples:         1024 | elapsed time per iteration (ms): 39375.6 | throughput per GPU (TFLOP/s/GPU): 310.5 | learning rate: 9.900000E-05 | global batch size:  1024 | lm loss: 2.404998E+00 | loss scale: 1.0 | grad norm: 3.048 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |Number of parameters in transformer block in billions:  6.98
 [2025-12-28 01:34:03] iteration        2/     100 | consumed samples:         2048 | elapsed time per iteration (ms): 35609.5 | throughput per GPU (TFLOP/s/GPU): 343.3 | learning rate: 9.800000E-05 | global batch size:  1024 | lm loss: 4.466912E+00 | loss scale: 1.0 | grad norm: 44.056 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
```

**Comment by BoxiangW** [comment]
Hi, we are actually trying to deprecate the usage of FSDP2 in MLM. It would be removed by next release

**Comment by haok1402** [comment]
@BoxiangW Perhaps adding a deprecation warning might be helpful?

**Comment by cspades** [comment]
Agreed with @BoxiangW, this `dist_ckpt` <> `FSDP2` checkpointing issue is likely going to be more effort to fix than to just deprecate instead. I can investigate out of curiosity what the current state of FSDP2 support in MLM is, but I plan to just add a deprecation warning and I'll also work on the FSDP2 deprecation as well.

Maybe we can bring back FSDP2 if we unify the Megatron checkpointing into DTensor / DCP, and then it'll be much easier to outsource checkpointing logic to native Torch.

**Comment by cspades** [comment]
@haok1402 Just a quick update, I am seeing a lot of users still using FSDP2 so I take back the deprecation comment from before. Need to find time to look into this, currently swamped with other tasks and in the near future.

Super aware this issue went in the wrong direction in what we should do about FSDP2, and it's a bazillion months old, just a priority thing, I'll RCA it later and update this issue then. No documentation either, written by someone who has left NVIDIA.

**Comment by cspades** [comment]
@haok1402 Okay, so this bug led me on a small goose chase but the reason the checkpoint breaks is because of SwiGLU, because of course it is.

Checkpoint loading happens after FSDP2 initialization, so we are looking at TP-DP strided shards in the model. However, `torch_dist` SwiGLU weights are saved like `[W; V]` and then during load redistributed as `[W_tp0 V_tp0 W_tp1 V_tp1 ... ]`. There's some code in `MLP.apply_swiglu_sharded_factory` that handles this redistribution. However, it doesn't know about FSDP2 strided sharding. (So basically, the DP-sharded `ShardedTensor` was swizzled into TP, when TP size was just 1 so no swizzling needed, corrupting the TP1 FSDP8 SwiGLU weight.)

I have a branch that supports redistributing all of the shards in `[ {W_tp0; V_tp0}_dp{1...DP}, {W_tp1; V_tp1}_dp{1...DP}, ... ]` to the correct ranks (read like, the DP ranks shard the concatenation of the TP-shards of W and V, which I'll admit kind of hurts to think about even after writing it out): https://github.com/cspades/Megatron-LM/tree/cye/fsdp2-ckpt

Not sure if it's mergeable yet (`is_dp_sharded = axis_frag == tp_size * dp_size and dp_size > 1` may not be a safe way to check for DP weight sharding... need to hack more FSDP2 flags 🤢 ) and still need to add DP>1 TP>1 checkpointing tests, but it should support MLP and ~~MoE GroupedLinear~~ (in theory, but EP is not supported for FSDP2 in MCore) with FSDP2. After this fix, we get activation and loss parity resuming from a pre-trained/non-trivial `torch_dist` checkpoint:
```
# FSDP
 [2026-07-07 14:44:43.435452] iteration       51/     100 | consumed samples:          816 | elapsed time per iteration (ms): 12756.0 | throughput per GPU (TFLOP/s/GPU): 15.0 | learning rate: 9.900000E-05 | global batch size:    16 | lm loss: 3.642244E-02 | loss scale: 1.0 | grad norm: 0.913 | number of skipped iterations:   0 | number of nan iterations:   0 |Number of parameters in transformer block in billions:  6.98

(Pdb) hidden_states
tensor([[[ 0.1367,  2.0469,  1.0938,  ..., -0.7500, -1.2734,  0.5156]],

        [[ 0.5586,  1.2656,  0.7148,  ..., -0.6406, -0.0315,  0.0825]],

        [[ 1.3672,  1.5312,  0.3164,  ...,  0.1377, -1.0156, -0.7812]],

        ...,

        [[-0.9375,  0.6797,  0.5664,  ..., -1.6406,  0.5352,  0.1543]],

        [[-0.2188, -0.7227, -0.0579,  ..., -1.9297, -1.2266, -1.0078]],

        [[ 2.0156,  1.1562,  1.3281,  ..., -0.1699, -0.3398, -2.4844]]],
       device='cuda:0', dtype=torch.bfloat16,
       grad_fn=<MakeViewlessTensorBackward>)

# ND Parallel
 [2026-07-07 14:43:32.368419] iteration       51/     100 | consumed samples:          816 | elapsed time per iteration (ms): 42941.8 | throughput per GPU (TFLOP/s/GPU): 4.4 | learning rate: 9.900000E-05 | global batch size:    16 | lm loss: 3.642244E-02 | loss scale: 1.0 | grad norm: 0.913 | number ofskipped iterations:   0 | number of nan iterations:   0 |Number of parameters in transformer block in billions:  6.98

(Pdb) hidden_states
tensor([[[ 0.1367,  2.0469,  1.0938,  ..., -0.7500, -1.2734,  0.5156]],

        [[ 0.5586,  1.2656,  0.7148,  ..., -0.6406, -0.0315,  0.0825]],

        [[ 1.3672,  1.5312,  0.3164,  ...,  0.1377, -1.0156, -0.7812]],

        ...,

        [[-0.9375,  0.6797,  0.5664,  ..., -1.6406,  0.5352,  0.1543]],

        [[-0.2188, -0.7227, -0.0579,  ..., -1.9297, -1.2266, -1.0078]],

        [[ 2.0156,  1.1562,  1.3281,  ..., -0.1699, -0.3398, -2.4844]]],
       device='cuda:0', dtype=torch.bfloat16,
       grad_fn=<MakeViewlessTensorBackward>)
```

I don't think this affects Megatron-FSDP since we have our own checkpointing format that has online handling of the W and V TP-shard and slicing out where we expect the DP-shard to land unevenly: [split_swiglu_linear_fc1](https://github.com/cspades/Megatron-LM/blob/f33e0096c65061ca4dacd652a5af36b09b65bbdb/megatron/core/transformer/fsdp_dtensor_checkpoint.py#L221)

**Comment by cspades** [comment]
Removing my script from my branch in preparation for main branch merge, here's my script, TL;DR don't load the distributed optimizer state if you trained with ND-parallel since FSDP2 doesn't use it (`TRAIN_CONFIG[no-load-optim]=true`):
```
#!/bin/bash

declare -A MODEL_CONFIG
MODEL_CONFIG[tokenizer-type]=HuggingFaceTokenizer
MODEL_CONFIG[tokenizer-model]=meta-llama/Meta-Llama-3-8B
MODEL_CONFIG[untie-embeddings-and-output-weights]=true
MODEL_CONFIG[position-embedding-type]=rope
MODEL_CONFIG[rotary-base]=500000
MODEL_CONFIG[max-position-embeddings]=8192
MODEL_CONFIG[num-layers]=32
MODEL_CONFIG[hidden-size]=4096
MODEL_CONFIG[ffn-hidden-size]=14336
MODEL_CONFIG[hidden-dropout]=0.0
MODEL_CONFIG[disable-bias-linear]=true
MODEL_CONFIG[swiglu]=true
MODEL_CONFIG[group-query-attention]=true
MODEL_CONFIG[num-attention-heads]=32
MODEL_CONFIG[num-query-groups]=8
MODEL_CONFIG[kv-channels]=128
MODEL_CONFIG[attention-dropout]=0.0
MODEL_CONFIG[init-method-std]=0.02
MODEL_CONFIG[normalization]=RMSNorm
MODEL_CONFIG[norm-epsilon]=1e-5

declare -A TRAIN_CONFIG
TRAIN_CONFIG[lr]=1e-4
TRAIN_CONFIG[optimizer]=adam
TRAIN_CONFIG[adam-beta1]=0.9
TRAIN_CONFIG[adam-beta2]=0.95
TRAIN_CONFIG[weight-decay]=0.1
TRAIN_CONFIG[clip-grad]=1.0
TRAIN_CONFIG[train-iters]=200
TRAIN_CONFIG[micro-batch-size]=1
TRAIN_CONFIG[global-batch-size]=16
TRAIN_CONFIG[seq-length]=2048
TRAIN_CONFIG[bf16]=true
TRAIN_CONFIG[mock-data]=true
TRAIN_CONFIG[vocab-size]=128256

# TRAIN_CONFIG[use-distributed-optimizer]=true
TRAIN_CONFIG[use-torch-fsdp2]=true
TRAIN_CONFIG[no-gradient-accumulation-fusion]=true

TRAIN_CONFIG[log-interval]=1
TRAIN_CONFIG[no-one-logger]=true
TRAIN_CONFIG[log-throughput]=true
TRAIN_CONFIG[load]=checkpoints/debug/megatron/  # torch_dist checkpoint -> FSDP2 model
TRAIN_CONFIG[save]=checkpoints/debug/megatron/
TRAIN_CONFIG[no-load-optim]=true
TRAIN_CONFIG[save-interval]=100
TRAIN_CONFIG[eval-interval]=250
TRAIN_CONFIG[eval-iters]=1

MAIN_ARGS=()
for key in ${!MODEL_CONFIG[@]}; do val=${MODEL_CONFIG[$key]}; [[ $val == true ]] && MAIN_ARGS+=(--$key) || MAIN_ARGS+=(--$key $val); done
for key in ${!TRAIN_CONFIG[@]}; do val=${TRAIN_CONFIG[$key]}; [[ $val == true ]] && MAIN_ARGS+=(--$key) || MAIN_ARGS+=(--$key $val); done

TRUN_ARGS=()
TRUN_ARGS+=(--nnodes=1 --nproc-per-node=8)
TRUN_ARGS+=(--rdzv-backend=c10d --rdzv-endpoint=localhost:15213)

SCRIPT=pretrain_gpt.py
STDOUT=checkpoints/debug/midtrain.log

mkdir -p $(dirname $STDOUT)
export WANDB_RUN_ID=$(echo ${TRAIN_CONFIG[WANDB_EXP_NAME]} | md5sum | cut -c1-5)
uv run python -m torch.distributed.run ${TRUN_ARGS[@]} $SCRIPT ${MAIN_ARGS[@]}
```

**Comment by haok1402** [comment]
@cspades thanks for digging into this and figuring out the root cause!! The SwiGLU strided-sharding thing makes sense. I've moved on to other stuff so I couldn't really offer anything insightful beyond the original thread, but yeah, the checkpoint conversion did feel pretty hacky when I was using it. Appreciate you tracking it down and glad to see the fix in #5714 :-)
